man dir does not change as specified by --mandir

<!-- Thank you for filing an issue to help us improve Dune! -->
## Problem

dune has been failing to build from source on **pkgsrc** because they can't get dune to put the man pages in a different location (`share/man/*`) from dune's default `man/*` (https://us-central.manta.mnx.io/pkgsrc/public/reports/upstream-trunk/20240305.1526/ocaml-dune-3.11.1/install.log).

As an experiment, I cloned dune from github to build it the make way. I executed the following sequence, from what I understood through various conversations on the github repo:

```
cd dune/
./configure --mandir=./fake_root/share/man
PREFIX=fake_root make release
PREFIX=fake_root make install
```

`configure` command seem to be doing as it is intended, and it creates/modifies `src/dune_rules/setup.ml`:

```
let library_path = []

let roots : string option Install.Roots.t =
  { lib_root = None
  ; man = Some "/Users/jb/projects/dune/./fake_root/share/man"
  ; doc_root = None
  ; etc_root = None
  ; share_root = None
  ; bin = None
  ; sbin = None
  ; libexec_root = None
  }
```

## Expected Behavior

man pages should be installed at `fake_root/share/man/*`.

## Actual Behavior

man pages are still installed at man/* instead of share/man/*.

```
> tree -d fake_root
fake_root
├── bin
├── doc
│   └── dune
│       └── odoc-pages
├── lib
│   └── dune
├── man
│   ├── man1
│   └── man5
└── share
    └── emacs
        └── site-lisp
```

## Reproduction

1. Clone dune repo
2. Execute the following sequence of commands.

```
cd dune/
./configure --mandir=./fake_root/share/man
PREFIX=fake_root make release
PREFIX=fake_root make install
```

## Specifications

- Version of `dune` (output of `dune --version`): `main` branch
- Version of `ocaml` (output of `ocamlc --version`): 4.14.1
- Operating system (distribution and version): macOS Sonoma 14.3 (Silicon aarm64 architecture)

> Instead of having `PREFIX`, doing:

On my end, the process fails without PREFIX, asking me to enter a prefix another way (`--prefix`).

```
$ rm -rf fake_root

$ make clean

$ ./configure --mandir=./fake_root/share/man

$ make release

$ make install
./_boot/dune.exe install    dune
Error: The docdir installation directory is unknown.
Hint: It can be specified with '--prefix' or by setting '--docdir'
make: *** [install] Error 1
```

--- a/boot/configure.ml
+++ b/boot/configure.ml
@@ -35,6 +35,7 @@ let toggle =
 
 let () =
   let bad fmt = ksprintf (fun s -> raise (Arg.Bad s)) fmt in
+  let prefix = ref None in
   let library_path = ref [] in
   let library_destdir = ref None in
   let mandir = ref None in
@@ -58,7 +59,8 @@ let () =
     v := Some dir
   in
   let args =
-    [ ( "--libdir"
+    [ "--prefix", Arg.String (set_dir prefix), "DIR where files are copied"
+    ; ( "--libdir"
       , Arg.String set_libdir
       , "DIR where public libraries are looked up in the default build context. Can be \
          specified multiple times new one taking precedence. The last is used as default \
@@ -120,6 +122,7 @@ let () =
   pr "  ; libexec_root = %s" (option string !libexecdir);
   pr "  }";
   pr "";
+  pr "let prefix : string option = %s" (option string !prefix);
   List.iter !toggles ~f:(fun (name, value) ->
     pr
       "let %s = `%s"

--- a/doc/changes/11172.md
+++ b/doc/changes/11172.md
@@ -1,0 +1,3 @@
+- Allow the `--prefix` flag when configuring dune with `ocaml configure.ml`.
+  This allows to set the prefix just like `$ dune install --prefix`. (#1172,
+  @rgrinberg)

--- a/src/dune_rules/context.ml
+++ b/src/dune_rules/context.ml
@@ -743,6 +743,19 @@ let map_exe (context : t) =
        | _ -> exe)
 ;;
 
+let roots =
+  lazy
+    (let open Setup in
+     match prefix with
+     | None -> roots
+     | Some prefix ->
+       let prefix = Install.Roots.make prefix ~relative:Filename.concat in
+       Install.Roots.map2 roots prefix ~f:(fun root prefix ->
+         match root with
+         | None -> Some prefix
+         | Some _ -> root))
+;;
+
 let roots t =
   let module Roots = Install.Roots in
   let+ prefix_roots =
@@ -756,7 +769,7 @@ let roots t =
   in
   match t.kind with
   | Lock _ | Default ->
-    let setup_roots = Roots.map ~f:(Option.map ~f:Path.of_string) Setup.roots in
+    let setup_roots = Roots.map ~f:(Option.map ~f:Path.of_string) (Lazy.force roots) in
     Roots.first_has_priority setup_roots prefix_roots
   | Opam _ -> prefix_roots
 ;;

--- a/src/dune_rules/setup.defaults.ml
+++ b/src/dune_rules/setup.defaults.ml
@@ -14,3 +14,4 @@ let roots : string option Install.Roots.t =
 let toolchains = `Enabled
 let pkg_build_progress = `Disabled
 let lock_dev_tool = `Disabled
+let prefix = None

--- a/src/dune_rules/setup.mli
+++ b/src/dune_rules/setup.mli
@@ -13,3 +13,4 @@ val roots : string option Install.Roots.t
 val toolchains : Dune_config.Config.Toggle.t
 val pkg_build_progress : Dune_config.Config.Toggle.t
 val lock_dev_tool : Dune_config.Config.Toggle.t
+val prefix : string option

--- a/src/install/roots.mli
+++ b/src/install/roots.mli
@@ -20,6 +20,7 @@ val opam_from_prefix : 'a -> relative:('a -> string -> 'a) -> 'a t
 val complete : 'a option t -> 'a option t
 
 val map : f:('a -> 'b) -> 'a t -> 'b t
+val map2 : f:('a -> 'b -> 'c) -> 'a t -> 'b t -> 'c t
 
 (** return the roots of the first argument if present *)
 val first_has_priority : 'a option t -> 'a option t -> 'a option t

